src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -112,9 +112,35 @@ public int Count
         
         public override int GetHashCode()
         {
-            return null == _array
-                        ? 0
-                        : _array.GetHashCode() ^ _offset ^ _count;
+            if (_array == null)
+            {
+                return 0;
+            }
+            
+#if !FEATURE_CORECLR
+            // Old implementation: Xor everything.
+            // This isn't optimal since it treats _offset and _count interchangeably,
+            // but we keep it around in the full framework for compat.
+            return _array.GetHashCode() ^ _offset ^ _count;
+#else
+            // Modified Bernstein hash: start with 5381,
+            // multiply by 33, xor with the next hash.
+            // NOTE: (hash << 5) + hash is equal to hash * 33.
+            
+            int hash = (33 * 5381) ^ _offset;
+            hash = ((hash << 5) + hash) ^ _count;
+            
+            int arrayHash = _array.GetHashCode();
+            // Note quite sure why, but it looks like the result of object.GetHashCode
+            // only returns 26 bits in the return value- the upper 6 are set to zero.
+            // See src/vm/syncblk.h for more.
+            // As a result, we take the array hash and shift it 6 to the left
```

Comment:
I do not think we need to make assumptions about object hash taking 26bits. 
It should be treated as a pseudo-random number not correlated with _offset or _count in any ways and as such it could be just directly mixed with a xor.